File: tree/bst.py
class TreeNode(object):
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None

def bst_insert(root, val):
    '''Insert value into BST. Note if value is already present, we're not
    inserting.  Note that below crisp code handles all cases.'''
    if root is None:
        return TreeNode(val)
    if val < root.val:
        root.left = bst_insert(root.left, val)
    elif val > root.val:
        root.right = bst_insert(root.right, val)
    return root

def bst_lowest(root):
    '''Find lowest value in BST'''
    if root is None:
        return
    curr = root
    while curr.left is not None:
        curr = curr.left
    return curr

def bst_highest(root):
    '''Find highest value in BST'''
    if root is None:
        return
    curr = root
    while curr.right is not None:
        curr = curr.right
    return curr

def bst_remove(root, val):
    '''Remove value from BST. The function returns value of the node which will
    replace current (i.e. root) node.'''
    if root is None:
        return None
    if root.val < val:
        root.right = bst_remove(root.right, val)
    elif root.val > val:
        root.left = bst_remove(root.left, val)
    else:
        if root.left is None:
            root = root.right
        elif root.right is None:
            root = root.left
        else:
            temp = bst_highest(root.left)
            root.val = temp.val
            root.left = bst_remove(root.left, temp.val)
    return root

def bst_print_inorder(root):
    if root is None:
        return
    bst_print_inorder(root.left)
    print(root.val)
    bst_print_inorder(root.right)

if __name__ == '__main__':
    root = TreeNode(50)
    bst_insert(root, 30)
    bst_insert(root, 20)
    bst_insert(root, 40)
    bst_insert(root, 70)
    bst_insert(root, 60)
    bst_insert(root, 80)
    bst_print_inorder(root)
    bst_remove(root, 50)
    print('=======')
    bst_print_inorder(root)
    print(root.val)
    print('done')

File: tree/bst.cpp
#include<iostream>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left, *right;
    TreeNode(int x): val(x), left(NULL), right(NULL) {};
};

TreeNode* insert(TreeNode *root, int val) {
    if (!root) {
        root = new TreeNode(val);
        return root;
    }
    if (root->val > val) {
        root->left = insert(root->left, val);
    } else if (root->val < val) {
        root->right = insert(root->right, val);
    }
    return root;
}

TreeNode* highest_node_lower_than_helper(TreeNode *root, TreeNode *curr, int val) {
    if (!root)
        return curr;
    if (!curr) {
        if (root->val < val) {
            curr = root;
        }
    } else if (root->val < val && root->val > curr->val) {
        curr = root;
    }

    if (root->val > val) {
        TreeNode* temp = highest_node_lower_than_helper(root->left, curr, val);
        return temp;
    } else if (root->val < val) {
        TreeNode* temp = highest_node_lower_than_helper(root->right, curr, val);
        return temp;
    } else {
        return root;
    }
}

// for given value, find node with larges value lower than or equal to given
// value
TreeNode* highest_node_lower_than(TreeNode *root, int val) {
    if (root->val <= val)
        return highest_node_lower_than_helper(root, root, val);
    return highest_node_lower_than_helper(root, NULL, val);
}

void kth_smallest_in_bst_helper(TreeNode *root, int k, int &i) {
    if (!root)
        return;
    kth_smallest_in_bst_helper(root->left, k, i);
    i++;
    if (i==k)
        cout<<root->val<<endl;
    kth_smallest_in_bst_helper(root->right, k, i);
}

void kth_smallest_in_bst(TreeNode *root, int k) {
    int i = 0;
    kth_smallest_in_bst_helper(root, k, i);
}


// int min_dist_to_leaf(TreeNode *root, int val) {
//     if (!root)
//         return 0;
//     return min(min_dist_to_leaf(root->right), min_dist_to_leaf(root->left));
// }


int main() {
    // tree nodes in inorder: 20 30 40 50 60 70 80
    TreeNode *root = new TreeNode(50);
    root->left = new TreeNode(30);
    root->left->left = new TreeNode(20);
    root->left->right = new TreeNode(40);
    root->right = new TreeNode(70);
    root->right->left = new TreeNode(60);
    root->right->right = new TreeNode(80);
    TreeNode *curr = highest_node_lower_than(root, 91);
    if (!curr)
        cout<<"no such element found"<<endl;
    else
        cout<<curr->val<<endl;
    kth_smallest_in_bst(root, 3);
    kth_smallest_in_bst(root, 1);
    kth_smallest_in_bst(root, 2);
    kth_smallest_in_bst(root, 4);
}
